### ISSUE TITLE
SD.h - Application is not notified that file::flush() failed

### ISSUE THREAD

**Issue by ChristopheLaurent** [espressif/arduino-esp32#6303]
### Related area

SD card

### Hardware specification

LilyGo ESP32 board with SD slot

### Is your feature request related to a problem?

My application can write data in a file on this SD card without any problem using `my_file.printf()`.
If I remove the SD card while writing, there is no error. I noticed that the size returned by `my_file.printf()` > 0 even if nothing was wrote on the SD card since the card was physically removed. The error occurs later when my application calls `my_file.flush()`. I can then see on serial monitor:
`[E][sd_diskio.cpp:123] sdSelectCard(): Select Failed`
I wish my application to react on this error instead of going on silently.
I did not find any public API in `SD.h` to inform my application that `my_file.flush()` raised an error.

### Describe the solution you'd like

I'd like SD class to keep track of such a failure in a `SD.h` public property.

### Additional context

I've seen issue SD card status check bug #6081, but I don't think this is related to my issue.

### I have checked existing list of Feature requests and the Contribution Guide

- [X] I confirm I have checked existing list of Feature requests and Contribution Guide.

**Comment by tobozo** [comment]
@ChristopheLaurent does this snippet from `sd_diskio.cpp` produce the expected behaviour?


```C++

bool sdSelectCard(uint8_t pdrv)
{
    ardu_sdcard_t * card = s_cards[pdrv];
    digitalWrite(card->ssPin, LOW);
    bool s = sdWait(pdrv, 300);
    static unsigned int failcount = 0;
    if (!s) {
		if (failcount == 0){
			log_e("Select Failed on first attempt!!!");
			//digitalWrite(card->ssPin, HIGH);
			//return false;
		} else {
			log_e("Select Failed");
			digitalWrite(card->ssPin, HIGH);
			return false;
		}
		failcount ++;
    }
    return true;
}
```

**Comment by P-R-O-C-H-Y** [comment]
Hi, can you please provide more informations that were typed in issue template?
What version of Arduino-ESP32 are you using?

**Comment by ChristopheLaurent** [comment]
Hi @P-R-O-C-H-Y , I was still using V1.0.6 according to file `/cores/esp32/core_version.h` and I realized that implementation of `sdSelectCard()` in `sd_diskio.cpp` changed since V1.0.6 . So I decided to move to V2 but it does not build with PlatformIO due to [known issue #619](https://github.com/platformio/platform-espressif32/issues/619) . Do you think that my original issue about SD card is already solved in V2 ? If not I prefer to stay on 1.0.6 until main V2 branch is fixed for PlatformIO.

**Comment by ChristopheLaurent** [comment]
Hi @tobozo, I guess it would help but I would prefer your `failcount` to be a private property of SD class with a public accessor instead of a static variable. Also I don't understand why we would need a special logic on your `failcount` to reset ss pin.

**Comment by P-R-O-C-H-Y** [comment]
> Hi @P-R-O-C-H-Y , I was still using V1.0.6 according to file `/cores/esp32/core_version.h` and I realized that implementation of `sdSelectCard()` in `sd_diskio.cpp` changed since V1.0.6 . So I decided to move to V2 but it does not build with PlatformIO due to [known issue #619](https://github.com/platformio/platform-espressif32/issues/619) . Do you think that my original issue about SD card is already solved in V2 ? If not I prefer to stay on 1.0.6 until main V2 branch is fixed for PlatformIO.

I don't know about issues with SD card on 1.0.6 version. But this version is no longer supported.
In 2.0.2 there were issues, which are already fixed in master branch. So for now is better for you to stay on 1.0.6 and wait until 2.0.3 is released (will be soon). But anyway if you are using platformio, Arduino-esp32 is not updated on their side. I will suggest you to follow some tutorial how to update Arduino-esp32 on yourself. I think its possible.
